src/core.jl

```diff
@@ -332,166 +332,122 @@ function jnew(T::Symbol, argtypes::Tuple = () , args...)
+function get_method_id(obj::JavaObject, method::AbstractString, rettype::Type, argtypes::Tuple)
     sig = method_signature(rettype, argtypes...)
-    jmethodId = JNI.GetMethodID(Ptr(metaclass(obj)), String(method), sig)
+    JNI.GetMethodID(Ptr(metaclass(obj)), String(method), sig)
+end
+
+function get_method_id(obj::JavaObject, method::JMethod)
+    sig = method_signature(rettype, argtypes...)
+    JNI.FromReflectedMethod(method)
+end
+
+function jcall(ref, method::JMethod, args...) where T
+    jmethodId = JNI.FromReflectedMethod(method)
+    rettype = jimport(getreturntype(method))
+    argtypes = Tuple(jimport.(getparametertypes(method)))
     jmethodId==C_NULL && geterror(true)
-    _jcall(obj, jmethodId, C_NULL, rettype,  argtypes, args...)
+    _jcall(metaclass(T), jmethodId, rettype, argtypes, args...)
 end
 
 function jcall(obj::JavaObject, method::JMethod, args... )
     assertroottask_or_goodenv() && assertloaded()
+    isnull(obj) && throw(JavaCallError("Attempt to call method on Java NULL"))
     jmethodId = JNI.FromReflectedMethod(method)
     rettype = jimport(getreturntype(method))
     argtypes = Tuple(jimport.(getparametertypes(method)))
     jmethodId==C_NULL && geterror(true)
-    _jcall(obj, jmethodId, C_NULL, rettype,  argtypes, args...)
+    _jcall(obj, jmethodId, rettype,  argtypes, args...)
 end
 
 # JMethod invoke
 (m::JMethod)(obj, args...) = jcall(obj, m, args...)
 
 
-function jfield(typ::Type{JavaObject{T}}, field::AbstractString, fieldType::Type) where T
+function jfield(ref, field, fieldType)
     assertroottask_or_goodenv() && assertloaded()
-    jfieldID  = JNI.GetStaticFieldID(Ptr(metaclass(T)), String(field), signature(fieldType))
+    jfieldID = get_field_id(ref, field, fieldType)
     jfieldID==C_NULL && geterror(true)
-    _jfield(metaclass(T), jfieldID, fieldType)
+    _jfield(_jcallable(ref), jfieldID, fieldType)
 end
 
-function jfield(obj::Type{JavaObject{T}}, field::JField) where T
-    assertroottask_or_goodenv() && assertloaded()
+function get_field_id(typ::Type{JavaObject{T}}, field::AbstractString, fieldType::Type) where T
+    JNI.GetStaticFieldID(Ptr(metaclass(T)), String(field), signature(fieldType))
+end
+
+function get_field_id(obj::Type{JavaObject{T}}, field::JField) where T
     fieldType = jimport(gettype(field))
-    jfieldID = JNI.FromReflectedField(field)
-    jfieldID==C_NULL && geterror(true)
-    _jfield(metaclass(T), jfieldID, fieldType)
+    JNI.FromReflectedField(field)
 end
 
-function jfield(obj::JavaObject, field::AbstractString, fieldType::Type)
-    assertroottask_or_goodenv() && assertloaded()
-    jfieldID  = JNI.GetFieldID(Ptr(metaclass(obj)), String(field), signature(fieldType))
-    jfieldID==C_NULL && geterror(true)
-    _jfield(obj, jfieldID, fieldType)
+function get_field_id(obj::JavaObject, field::AbstractString, fieldType::Type)
+    JNI.GetFieldID(Ptr(metaclass(obj)), String(field), signature(fieldType))
 end
 
-function jfield(obj::JavaObject, field::JField)
```

Comment:
Did we drop the two argument version of `jfield`?

Reply:
Good catch! I'm expanding the test suite to test for this as well as some other things.